File: project-java-cosminpac04/motorcycleContest/src/main/java/motorcycle/repository/UserRepository.java
package motorcycle.repository;

import motorcycle.database.DataBaseConnection;
import motorcycle.model.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class UserRepository implements IRepository<User, Integer> {
    private static final Logger logger = LoggerFactory.getLogger(UserRepository.class);

    @Override
    public void add(User user) {
        String sql = "INSERT INTO users (Uid, UserName, Password, Role) VALUES (?, ?, ?)";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, user.getID());
            stmt.setString(2, user.getUsername());
            stmt.setString(3, user.getPassword());
            stmt.setString(4, user.getRole());
            stmt.executeUpdate();

            logger.info("Added user: {}", user.getUsername());

        } catch (SQLException e) {
            logger.error("Error adding user: {}", user.getUsername(), e);
        }
    }

    @Override
    public void delete(User user) {
        String sql = "DELETE FROM users WHERE Uid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, user.getID());
            stmt.executeUpdate();

            logger.info("Deleted user with ID: {}", user.getID());

        } catch (SQLException e) {
            logger.error("Error deleting user with ID: {}", user.getID(), e);
        }
    }

    @Override
    public void update(User user, Integer id) {
        String sql = "UPDATE users SET UserName = ?, Password = ?, Role = ? WHERE Uid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, user.getUsername());
            stmt.setString(2, user.getPassword());
            stmt.setString(3, user.getRole());
            stmt.setInt(4, id);
            stmt.executeUpdate();

            logger.info("Updated user with ID: {}", id);

        } catch (SQLException e) {
            logger.error("Error updating user with ID: {}", id, e);
        }
    }

    @Override
    public User findById(Integer id) {
        String sql = "SELECT * FROM users WHERE Uid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, id);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                User user = new User(rs.getInt("id"), rs.getString("username"), rs.getString("password"), rs.getString("role"));
                logger.info("Found user: {}", user.getUsername());
                return user;
            } else {
                logger.warn("No user found with ID: {}", id);
            }

        } catch (SQLException e) {
            logger.error("Error finding user with ID: {}", id, e);
        }
        return null;
    }

    @Override
    public List<User> findAll() {
        List<User> users = new ArrayList<>();
        String sql = "SELECT * FROM users";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {

            while (rs.next()) {
                users.add(new User(rs.getInt("id"), rs.getString("username"), rs.getString("password"), rs.getString("role")));
            }

            logger.info("Retrieved {} users", users.size());

        } catch (SQLException e) {
            logger.error("Error retrieving users", e);
        }
        return users;
    }

    @Override
    public List<User> getAll() {
        return findAll();
    }
}

File: project-c-cosminpac04/motorcycleApp/Repositories/UserRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using motorcycleApp.DataBase;
using motorcycleApp.Models;
using Serilog;

namespace motorcycleApp.Repositories;

public class UserRepository : IRepository<User, int>
{
    public void Add(User entity)
    {
        string sql = "INSERT INTO User (Username, Password, Role) VALUES (@Username, @Password, @Role)";
        try
        {
            using (var conn = ConnectionManager.GetConnection())
            {
                Log.Information("Connection established: " + conn.Database);

                using (var command = new SQLiteCommand(sql, conn))
                {
                    command.Parameters.AddWithValue("@Username", entity.Username);
                    command.Parameters.AddWithValue("@Password", entity.Password);
                    command.Parameters.AddWithValue("@Role", entity.Role);

                    command.ExecuteNonQuery();
                    Log.Information("Added user '{Username}' to Database", entity.Username);
                }
            }
        }
        catch (Exception ex)
        {
            Log.Error(ex, "Error while adding user '{Username}' to Database", entity.Username);
        }
    }

    public void Update(User entity)
    {
        string sql = "UPDATE User SET Username=@Username, Password=@Password, Role=@Role WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            {
                command.Parameters.AddWithValue("@Id", entity.ID);
                command.Parameters.AddWithValue("@Username", entity.Username);
                command.Parameters.AddWithValue("@Password", entity.Password);
                command.Parameters.AddWithValue("@Role", entity.Role);

                command.ExecuteNonQuery();
                Log.Information("Updated user ID {UserId} to '{Username}'", entity.ID, entity.Username);
            }
        }
    }

    public void Delete(User entity)
    {
        string sql = "DELETE FROM User WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            {
                command.Parameters.AddWithValue("@Id", entity.ID);

                int affectedRows = command.ExecuteNonQuery();
                if (affectedRows > 0)
                {
                    Log.Information("Deleted user ID {UserId} from Database", entity.ID);
                }
                else
                {
                    Log.Warning("No user found to delete with ID {UserId}", entity.ID);
                }
            }
        }
    }

    public User FindById(int id)
    {
        string sql = "SELECT * FROM User WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            {
                command.Parameters.AddWithValue("@Id", id);
                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        var user = new User(
                            reader.GetInt32(0),    // ID
                            reader.GetString(1),   // Username
                            reader.GetString(2),   // Password
                            reader.GetString(3)    // Role
                        );
                        Log.Information("Found user with ID {UserId}: '{Username}'", user.ID, user.Username);
                        return user;
                    }
                }
            }
        }

        Log.Warning("Could not find user with ID {UserId}", id);
        return null;
    }

    public IEnumerable<User> GetAll()
    {
        string sql = "SELECT * FROM User";
        var users = new List<User>();

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            using (var reader = command.ExecuteReader())
            {
                while (reader.Read())
                {
                    users.Add(new User(
                        reader.GetInt32(0),
                        reader.GetString(1),
                        reader.GetString(2),
                        reader.GetString(3)
                    ));
                }
            }
        }

        Log.Information("Retrieved {Count} users from Database", users.Count);
        return users;
    }
}
